sui_programmability/framework/sources/Governance/ValidatorSet.move

```diff
@@ -0,0 +1,296 @@
+        assert!(found, EVALIDATOR_NOT_FOUND);
+        assert!(
+            !Vector::contains(&self.pending_withdraws, &validator_index),
+            EDUPLICATE_WITHDRAW
+        );
+        Vector::push_back(&mut self.pending_withdraws, validator_index);
+    }
+
+    /// Called by `SuiSystem`, to add more stake to a validator.
+    /// If the validator is a pending one, we add stake to it directly; otherwise we send a request
+    /// to add the extra stake at the end of epoch.
+    public(friend) fun request_add_stake(
+        self: &mut ValidatorSet,
+        new_stake: Coin<SUI>,
+        validator_address: address,
+    ) {
+        let (found, validator_index) = find_validator(&self.pending_validators, validator_address);
+        if (found) {
+            let validator = Vector::borrow_mut(&mut self.pending_validators, validator_index);
+            Validator::add_stake_to_pending_validator(validator, new_stake);
+        } else {
+            let (found, validator_index) = find_validator(&self.active_validators, validator_address);
+            assert!(found, EVALIDATOR_NOT_FOUND);
+            let validator = Vector::borrow_mut(&mut self.active_validators, validator_index);
+            Validator::request_add_stake_to_active_validator(validator, new_stake);
+        }
+    }
+
+    /// Called by `SuiSystem`, to withdraw stake from a validator.
+    /// If the validator is a pending one, we withdraw from it and send back the coins immediately;
+    /// otherwise we send a withdraw request which will be processed at the end of epoch.
+    public(friend) fun request_withdraw_stake(
+        self: &mut ValidatorSet,
+        withdraw_amount: u64,
+        ctx: &mut TxContext,
+    ) {
+        let validator_address = TxContext::sender(ctx);
+        let (found, validator_index) = find_validator(&self.pending_validators, validator_address);
+        if (found) {
+            let validator = Vector::borrow_mut(&mut self.pending_validators, validator_index);
+            Validator::withdraw_stake_from_pending_validator(validator, withdraw_amount, ctx);
+        } else {
+            let (found, validator_index) = find_validator(&self.active_validators, validator_address);
+            assert!(found, EVALIDATOR_NOT_FOUND);
+            let validator = Vector::borrow_mut(&mut self.active_validators, validator_index);
+            Validator::request_withdraw_stake_from_active_validator(validator, withdraw_amount);
+        }
+    }
+
+    /// Update the validator set at the end of epoch.
+    /// It does the following things:
+    ///   1. Distribute stake award.
+    ///   2. Process pending stake deposits and withdraws for each validator (`adjust_stake`).
+    ///   3. Process pending validator application and withdraws.
+    ///   4. At the end, we decide which validators will form the validator set for the next epoch.
+    public(friend) fun advance_epoch(
+        self: &mut ValidatorSet,
+        new_validator_set_size: u64,
```

Comment:
I think validator set size and min/max stake are not things that will change on an epoch-by-epoch basis. Any change would probably be part of a protocol upgrade + would (probably) take effect by publishing by updating `SystemParameters`?